app/waf/test/acceptance/application.controller.acceptance.ts

```diff
@@ -264,4 +231,104 @@ describe('ApplicationController', () => {
       .send(application)
       .expect(404);
   });
+
+  it(
+    'post ' +
+      prefix +
+      '/applications/{id}/deploy: no ADC associated with application',
+    async () => {
+      const application = await givenApplicationData(wafapp);
+
+      await client
+        .post(prefix + '/applications/' + application.id + '/deploy')
+        .expect(422);
+    },
+  );
+
+  it(
+    'post ' + prefix + '/applications/{id}/deploy: application has no service',
+    async () => {
+      const application = await givenApplicationData(wafapp);
+      const adc = await givenAdcData(wafapp);
+      await givenTenantAssociationData(wafapp, {
+        tenantId: 'default',
+        adcId: adc.id,
+      });
+
+      await client
+        .post(prefix + '/applications/' + application.id + '/deploy')
+        .expect(422);
+    },
+  );
+
+  it(
+    'post ' + prefix + '/applications/{id}/deploy: deploy as3 config',
+    async () => {
+      const adc = await givenAdcData(wafapp);
+      await givenTenantAssociationData(wafapp, {
+        tenantId: 'default',
+        adcId: adc.id,
+      });
+      let member = await givenMemberData(wafapp);
+      let pool = await givenPoolData(wafapp, {
+        name: 'pool1',
+        members: [member.id],
+      });
+      let service = await givenServiceData(wafapp, {
+        pool: pool.id,
+      });
+      let application = await givenApplicationData(wafapp, {
+        services: [service.id],
+      });
+
+      let s = sinon
+        .stub(controller.as3Service, 'deploy')
+        .returns(Promise.resolve('Hello'));
+
+      await client
+        .post(prefix + '/applications/' + application.id + '/deploy')
+        .expect(200);
+
+      // Just intend to increase code coverage
+      let req = <AS3DeployRequest>s.getCall(0).args[2];
+      req.declaration.toJSON();
+
+      s.restore();
+    },
+  );
+
+  it(
+    'post ' + prefix + '/applications/{id}/deploy: no member in pool',
+    async () => {
+      const adc = await givenAdcData(wafapp);
+      await givenTenantAssociationData(wafapp, {
+        tenantId: 'default',
+        adcId: adc.id,
+      });
+      let pool = await givenPoolData(wafapp, {
+        name: 'pool1',
+        members: [],
+      });
+      let service = await givenServiceData(wafapp, {
+        pool: pool.id,
+      });
+      let application = await givenApplicationData(wafapp, {
+        services: [service.id],
+      });
+
+      let s = sinon
+        .stub(controller.as3Service, 'deploy')
+        .returns(Promise.resolve('Hello'));
+
+      await client
+        .post(prefix + '/applications/' + application.id + '/deploy')
+        .expect(200);
+
+      // Just intend to increase code coverage
+      let req = <AS3DeployRequest>s.getCall(0).args[2];
+      req.declaration.toJSON();
+
+      s.restore();
```

Comment:
Move to `before` and `after`.

Reply:
Not all case need this.

Reply:
There are many test cases trigger it. We should move it over just like `emptyDatabase` is not required for all test cases, but it's still be encapsulated into setup/teardown.